File: python/grpc_demo/list_movies.py
#! /usr/bin/env python
"""Iterate through all movies and print info about them"""

import sys
import movie_catalogue_pb2


def list_movies(movie_catalogue):
    """Iterate through all movies and print info about them"""
    for movie in movie_catalogue.movies:
        print("Title:", movie.title)
        print("\tDescription:", movie.description)
        print("\tProduction Year:", movie.productionYear)
        print("\tGenre:", movie.genre)
        print("\tDuration:", movie.duration)

        for cast_member in movie.castMembers:
            print("\t\tCharacter:", cast_member.character)
            print("\t\tFirst Name:", cast_member.firstName)
            print("\t\tLast Name:", cast_member.lastName)


def usage():
    print("Usage:", sys.argv[0], "MOVIE_CATALOGUE_FILE")
    sys.exit(-1)


def main():
    """Read the entire movie catalogue from a file and print all the
       information inside"""
    if len(sys.argv) != 2:
        usage()

    movie_catalogue = movie_catalogue_pb2.MovieCatalogue()

    # Read the existing movie catalogue
    with open(sys.argv[1], "rb") as f:
        movie_catalogue.ParseFromString(f.read())

    list_movies(movie_catalogue)


if __name__ == "__main__":
    main()

File: go/grpc_demo/list_movies.go
package main

import (
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"os"

	"github.com/golang/protobuf/proto"
)

func writeMovie(w io.Writer, m *Movie) {
	fmt.Fprintln(w, "Title:", m.Title)
	fmt.Fprintln(w, "  Description:", m.Description)
	fmt.Fprintln(w, "  Production Year:", m.ProductionYear)
	fmt.Fprintln(w, "  Genre:", m.Genre)
	fmt.Fprintln(w, "  Duration:", m.Duration)

	for _, cm := range m.CastMembers {
		fmt.Fprintln(w, cm.Character)
		fmt.Fprintln(w, cm.FirstName)
		fmt.Fprintln(w, cm.LastName)
	}
}

func listMovies(w io.Writer, movieCatalogue *MovieCatalogue) {
	for _, m := range movieCatalogue.Movies {
		writeMovie(w, m)
	}
}

// Main reads the entire address book from a file and prints all the
// information inside.
func main() {
	if len(os.Args) != 2 {
		log.Fatalf("Usage:  %s MOVIE_CATALOGUE_FILE\n", os.Args[0])
	}
	fname := os.Args[1]

	// [START unmarshal_proto]
	// Read the existing address book.
	in, err := ioutil.ReadFile(fname)
	if err != nil {
		log.Fatalln("Error reading file:", err)
	}
	movieCatalogue := &MovieCatalogue{}
	if err := proto.Unmarshal(in, movieCatalogue); err != nil {
		log.Fatalln("Failed to parse movie catalogue:", err)
	}
	// [END unmarshal_proto]

	listMovies(os.Stdout, movieCatalogue)
}
